Issue 712

synkarius opened the pull request:
# Fix module/package name conflict bug

## Description

The `ContentRequestGenerator` can now load modules with packages of the same name.

## How Has This Been Tested

Manual testing + added unit test.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue or bug)

## Checklist

- [ ] I have read the CONTRIBUTING document.
- [ ] My code follows the code style of this project.
- [ ] I have checked that my code does not duplicate functionality elsewhere in Caster.
- [ ] I have checked for and utilized existing command phrases from within Caster (delete if not applicable). 
- [x] My code implements all the features I wish to merge in this pull request.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests pass.

## Maintainer/Reviewer Checklist

<!-- Please leave these unchecked and add any other specific tasks you would like a -->
<!-- reviewer or maintainer to complete. -->

- [ ] Basic functionality has been tested and works as claimed.
- [x] New documentation is clear and complete.
- [x] Code is clear and readable.

LexiconCode commented:
Related issue but different underlying issue. Test procedure:

1. `pip install markdown`
2. restart caster
3. enable markdown
4. commands not recognized which is solved by uninstalling the markdown package.

mpourmpoulis commented:
Thanks for taking a look at this @synkarius Unfortunately I fear your solution relies among others on the assumption that the grandparent directory of The file involved is inside the `sys.path` So unles The Grandparent directory contains a grammar Or  is  the user directory itself I think the problem should persist

Also as demonstrated by @LexiconCode Because we are [Appending to the seas path](https://github.com/synkarius/Caster/blob/2d11819958dfd4a35481c20b1fc856ef49d9392b/castervoice/lib/ctrl/mgr/loading/load/content_loader.py#L24-L34) Site packages take precedence

synkarius commented:
@LexiconCode, @mpourmpoulis -- thank you for the additional examples and insights. That first solution was indeed incomplete.

For this fix, I moved a few things in the Caster user dir, and added a few things. (I moved the rules, hooks, and transformers dirs into a "caster_user_content" dir, and added __init__.py files to all of them.) With this, I added a migrator util, so no one should have to adjust anything manually. Any files they have in the old structure will be moved to the new structure the next time they start Caster.

I need to update the docs for that part, and there are may be a few more tests I'd like to write before we merge this, but it should be ready for testing again.

synkarius commented:
I added that test I was thinking of and updated the docs. I _think_ I got everything in the docs, but you might want to take a quick look at that last commit I pushed and see if there's anything I obviously missed.

LexiconCode commented:
hmm [bringme.py](https://github.com/dictation-toolbox/Caster/blob/master/castervoice/rules/ccr/recording_rules/bringme.py) will have to be updated with the new paths

However migrating user the generated toml might be a bit more complicated. Be aware if there is a syntax error that it reverts to the default and wiping out settings.

LexiconCode commented on castervoice/lib/settings.py:
```diff
@@ -496,9 +496,11 @@ def initialize():
     _SETTINGS_PATH = str(Path(_USER_DIR).joinpath("settings/settings.toml"))
 
-    for directory in ["data", "rules", "transformers", "hooks", "sikuli", "settings"]:
+    for directory in ["data", "sikuli", "settings"]:
```
We might as well bundle the rest of this into update_user_dir_packages_to_v1_7_0. Then it can be to simply update user directory. Beyond version migration I suppose if someone chooses to do a custom user directory path this could be reused migrate the files.

synkarius replied:
I agree. I'll move this stuff into the migrator util as well.

synkarius commented:
Nice catch with the `bringme.py` paths and toml. I think migrating the config can be done safely (by thoroughly unit testing it), but I can also create a `sm_bringme.toml.bak` file if you think it would put people's minds at ease.

LexiconCode commented:
> ), but I can also create a `sm_bringme.toml.bak` file if you think it would put people's minds at ease.

Yeah that would be pretty simple to do.

LexiconCode commented on castervoice/lib/ctrl/mgr/rule_details.py:
```diff
@@ -38,8 +38,8 @@ def __init__(self, name=None, function_context=None, executable=None, title=None
-            module = inspect.getmodule(frame[0])
-            filepath = module.__file__.replace("\\", "/")  
+            # module = inspect.getmodule(frame[1])
+            filepath = frame[1].replace("\\", "/")
```
Would it be better to let Path lib handle file path translation?

synkarius replied:
It would. :)

synkarius commented:
That should cover everything discussed above.

LexiconCode commented:
Bug report from chat

I deleted the user directory and ran the file . Did not work. Here is what I get. Running WRS from Dragonfly CLI.
INFO:engine:Initialized 'sapi5inproc' SR engine: Sapi5InProcEngine().
INFO:engine:Selecting audio source: 'Microphone (FIFINE K670 Microphone)'
INFO:module:CommandModule('_caster.py'): Loading module: 'C:\Users\capatou\Documents\Caster\_caster.py'


 FileNotFoundError(2, 'No such file or directory') while loading settings file: C:\Users\capatou\AppData\Local\caster\settings\settings.toml
Attempting to recover...


Default settings values added: 11
Error saving toml file: [Errno 2] No such file or directory: 'C:\\Users\\capatou\\AppData\\Local\\caster\\settings\\settings.toml' C:\Users\capatou\AppData\Local\caster\settings\settings.toml
ERROR:module:CommandModule('_caster.py'): Error loading module: 'NoneType' object is not subscriptable
Traceback (most recent call last):
  File "C:\Users\capatou\AppData\Local\Programs\Python\Python38\lib\site-packages\dragonfly\loader.py", line 65, in load
    exec(compile(contents, self._path, 'exec'), namespace)
  File "C:\Users\capatou\Documents\Caster\_caster.py", line 17, in <module>
    settings.initialize()
  File "C:\Users\capatou\Documents\Caster\castervoice\lib\settings.py", line 505, in initialize
    migrator.update_bringme_toml_to_v1_7_0()
  File "C:\Users\capatou\Documents\Caster\castervoice\lib\migration.py", line 42, in update_bringme_toml_to_v1_7_0
    if ContentRoot.USER_DIR not in directory_dict["caster rules"]:
TypeError: 'NoneType' object is not subscriptable

LexiconCode commented:
I think it's just a path error  "C:\Users\capatou\Documents\Caster_caster.py", line 17, in
settings.initialize(). I can fix it some time today.

LexiconCode commented on castervoice/lib/migration.py:
```diff
@@ -0,0 +1,79 @@
+        bringme_config.load()
+        directory_dict = bringme_config.get("folder")
+
+        if ContentRoot.USER_DIR not in directory_dict["caster rules"]:
```
`if directory_dict is not None and ContentRoot.USER_DIR not in directory_dict["caster rules"]:`

checking directory_dict if is None which would be the case if the user directory hasn't been generated yet on 1st run.

LexiconCode replied:
https://github.com/dictation-toolbox/Caster/commit/0470ec9ed58a3a64f9652c331868c30b108b9be5
